includes/class-wc-payments-payment-request-button-handler.php

```diff
@@ -104,6 +104,95 @@ public function init() {
 		// will be used to calculate it whenever the option value is retrieved instead.
 		// It's used for displaying inbox notifications.
 		add_filter( 'pre_option_wcpay_is_apple_pay_enabled', [ $this, 'get_option_is_apple_pay_enabled' ], 10, 1 );
+
+		if ( WC_Payments_Features::is_tokenized_cart_prb_enabled() ) {
+			add_filter( 'rest_pre_dispatch', [ $this, 'tokenized_cart_store_api_address_normalization' ], 10, 3 );
+		}
+	}
+
+	/**
+	 * Google Pay/Apple Pay parameters for address data might need some massaging for some of the countries.
+	 * Ensuring that the Store API doesn't throw a `rest_invalid_param` error message for some of those scenarios.
+	 *
+	 * @param mixed            $response Response to replace the requested version with.
+	 * @param \WP_REST_Server  $server Server instance.
+	 * @param \WP_REST_Request $request Request used to generate the response.
+	 *
+	 * @return mixed
+	 */
+	public function tokenized_cart_store_api_address_normalization( $response, $server, $request ) {
+		if ( 'true' !== $request->get_header( 'X-WC-Payments-prb-request' ) ) {
+			return $response;
+		}
+
+		// This route is used to get shipping rates.
+		// GooglePay/ApplePay might provide us with "trimmed" zip codes.
+		// If that's the case, let's temporarily allow to skip the zip code validation, in order to get some shipping rates.
+		if ( $request->get_route() === '/wc/store/v1/cart/update-customer' ) {
+			add_filter( 'woocommerce_validate_postcode', [ $this, 'maybe_skip_postcode_validation' ], 10, 3 );
```

Comment:
I tried to be conservative by ensuring that the filter only runs when the `X-WooPayments-Express-Payment-Request` header is added and when the `/wc/store/v1/cart/update-customer` request is performed.

I tried to be more specific with the validation, using a zip code that ends with a different value than `0` 🤷 
I wasn't able to find a better solution, but it sounds like partial address matching might be something that could be implemented in the future: p1714466491520799-slack-C02TS23QJ1X

Reply:
> I tried to be conservative by ensuring that the filter only runs when the X-WooPayments-Express-Payment-Request header is added and when the /wc/store/v1/cart/update-customer request is performed.

If we only want this filter to be run in a single use-case, don't we then need to remove the filter as well at some point pretty shortly after? I think it's fine on the PDP page, where there is no other way to run through WC checkout validation AFAIA, but if this implementation was present on the checkout page, for example, could I begin the express checkout flow, to effect this filter being added via the AJAX request. If I then exited the express checkout flow and began a regular checkout on the same page, would this filter now be present and would I be able to potentially bypass the usual WC validation? 🤔 

Or am I more simply fundamentally misunderstanding how filters work? 😆

Reply:
> but if this implementation was present on the checkout page, for example, could I begin the express checkout flow, to effect this filter being added via the AJAX request

Correct, this filter would be present on cart/checkout pages if the customer started the express checkout flow - but it would only execute on express checkout requests (which include the `X-WooPayments-Express-Payment-Request` header).
It is needed because the express checkout flow could provide partially anonymized address data both on PDPs and cart/checkout.

> If I then exited the express checkout flow and began a regular checkout on the same page, would this filter now be present and would I be able to potentially bypass the usual WC validation? 🤔

If you exited out of the express checkout flow, subsequent requests would not have this filter applied.
On each request to the `/wc/store/v1/cart/update-customer` endpoint of the Store API, this filter checks if the `X-WC-Payments-prb-request` is present or not.
The "skip validation" step is not executed if the header is absent and the request URL doesn't match `/wc/store/v1/cart/update-customer`.

Reply:
Cool, since you've explored it already with the blocks team and this was their recommendation, I'll accept your wiser words. Someone could feasibly inject this header into other requests on the blocks checkout, but there's presumably pretty limited damage that could be wrought by evading postcode validation on this particular endpoint. 👍 

I know that WP nonces are generally pretty feeble forms of defence against invader attacks, but do you think it might be helpful to add a nonce check to this function before applying the filter or is this particularly impotent if the nonce will be checked further upstream by the WC checkout anyways? 🤔

Reply:
The Store API already verifies and requires a nonce to be present (unless nonce checks are disabled by the `woocommerce_store_api_disable_nonce_check` 😬 ).
The Store API nonce check occurs after the `rest_pre_dispatch` filter, but before the API does anything with the data (or the filter).

I'm leaning two possible ways - we could either:
- Duplicate the nonce check from the Store API (checking the `wc_store_api` nonce in case nonce checks are disabled via the `woocommerce_store_api_disable_nonce_check` filter)
- Create a brand new WooPayments nonce to add to the request and be verified in this scenario

Do you lean in any specific way?

Reply:
Since this hook will fire and intercept every single API request and is essentially a public endpoint, I, personally, am leaning on the side of extreme--and likely excessive--caution. Right now, the only verification for this endpoint is the header and route path, neither of which need necessarily originate from our client and either of which is trivial to include in a request. Currently, I can't see that there is any particularly malicious crime that any third-party might be able to commit with the current functionality of this endpoint; but that doesn't necessarily prove that this endpoint is entirely harmless, nor that we might not carelessly add something to this endpoint in the future that could make it more vulnerable.

I think relying on the Store API to verify the nonce is also a little presumptuous, since it can apparently be disabled with a filter and moreover relying on security from a plugin that we do not necessarily control also feels a little risky. Personally, I would probably create a new nonce and verify it ourselves, just for an extra layer of security: this imitates the similar layer of possibly excessive security that we placed upon our [UPE redirection payment endpoint](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payment-gateway-wcpay.php#L1821), which also resides on top of a very public hook. 

Again, I don't think this endpoint is particularly vulnerable now, but I thought the same about our old UPE endpoint and we all know how that story ended. It seems like a low effort and low cost layer of over-protective security that could be beneficial in the long-run. 🤷

Reply:
That's fair enough - I added an additional nonce & nonce check in c6ea53133.

I thought I could just add the nonce as a parameter in `wcpayPaymentRequestParams`, but that didn't work for anonymous carts.
So I added an additional header to Store API responses.
This is because nonces are tied to the user session, and the anonymous cart counts as a separate user session.